abdiff/core/calc_ab_metrics.py

```diff
@@ -0,0 +1,291 @@
+    were no changes to them.
+    """
+    # create view of record diff matrix
+    conn.execute(
+        f"""
+            create view record_diff_matrix as (
+            select * from '{os.path.abspath(field_matrix_parquet)}'
+            );"""
+    )
+
+    # create table of field names with changes
+    conn.execute(
+        f"""
+            create table record_fields as (
+                select column_name as field_name
+                from information_schema.columns
+                where table_name = 'record_diff_matrix'
+                and column_name not in {tuple(NON_TIMDEX_FIELD_COLUMNS)}
+            );
+            """
+    )
+
+    # get list of TIMDEX fields that had changes in at least 1+ records
+    columns = conn.execute("PRAGMA table_info('record_diff_matrix');").fetchall()
+    fields = [col[1] for col in columns if col[1] not in NON_TIMDEX_FIELD_COLUMNS]
+
+    # get list of unique sources from the run
+    sources = [
+        row[0]
+        for row in conn.execute(
+            "select distinct source from record_diff_matrix;"
+        ).fetchall()
+    ]
+
+    return fields, sources
+
+
+def _get_global_counts(conn: DuckDBPyConnection, fields: list[str]) -> tuple[int, int]:
+    total_records = conn.execute(
+        """
+        SELECT COUNT(*) FROM record_diff_matrix
+        """
+    ).fetchone()[0]
+
+    any_field_modified_condition = " OR ".join(f"{field} = 1" for field in fields)
+    total_records_with_diff = conn.execute(
+        f"""
+        select count(*) from record_diff_matrix where {any_field_modified_condition}
+        """
+    ).fetchone()[0]
+
+    return total_records, total_records_with_diff
+
+
+def _get_source_counts(
+    conn: DuckDBPyConnection,
+    fields: list[str],
+    sources: list[str],
+    analysis: dict,
+) -> dict:
+    any_field_modified_condition = " OR ".join(f"{field} = 1" for field in fields)
+    for source in sources:
+        total_count = conn.execute(
+            f"""
+            select count(*) from record_diff_matrix
+            where source = '{source}' and ({any_field_modified_condition})
+        """
+        ).fetchone()[0]
+
+        source_counts = {"count": total_count, "field_counts": {}}
+
+        for field in fields:
+            field_count = conn.execute(
+                f"""
+                select count(*) from record_diff_matrix
+                where source = '{source}' and {field} = 1
+            """
+            ).fetchone()[0]
+            source_counts["field_counts"][field] = field_count
+
+        analysis["by_source"][source] = source_counts
+
+    return analysis
+
+
+def _get_field_counts(
+    conn: DuckDBPyConnection,
+    fields: list[str],
+    sources: list[str],
+    analysis: dict,
+) -> dict:
+    for field in fields:
+        total_field_count = conn.execute(
+            f"""
+            select count(*) from record_diff_matrix
+            where {field} = 1
```

Comment:
Bit of an arbitary comment, observing that DuckDB becomes really helpful for querying the sparse matrix of rows vs TIMDEX fields, where the presence of boolean `1` indicates that record, for that field, had a diff.  With all data at-rest as parquet, these DuckDB queries are very quick and memory efficient.